File: examples/python/ex03_llm_chat/main.py
"""03 LLM Chat — two LLM chat graphs.

Requires ``OPENAI_API_KEY`` in ``.env`` and ``llm:gpt-4o-mini`` in
``resources.yaml``. Run from this directory:

    uv sync
    cp .env.example .env  # fill in OPENAI_API_KEY
    uv run python main.py
"""

from __future__ import annotations

import asyncio

import operonx
from operonx.core import END, START, Operon, graph, op
from operonx.providers import LLMOp


@op
def clean_text(text: str):
    return {"cleaned_text": " ".join(text.split()).strip()}


@graph
def basic_chat(question):
    """Single-op form: LLMOp formats the prompt and calls the model."""
    llm = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={
            "system": "Bạn là trợ lý AI thân thiện. Trả lời ngắn gọn.",
            "user": "{question}",
        },
        question=question,
    )
    START >> llm >> END


@graph
def summarize_pipeline(text):
    """Pre-process → LLMOp with template."""
    pre = clean_text(text=text)
    summarize = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={
            "system": "Bạn là chuyên gia tóm tắt văn bản. Tóm tắt ngắn gọn trong 1-2 câu.",
            "user": "Tóm tắt:\n\n{text}",
        },
        text=pre["cleaned_text"],
    )
    START >> pre >> summarize >> END


SAMPLE_TEXT = (
    "        Trí tuệ nhân tạo (AI) đang thay đổi cách chúng ta sống và làm việc.\n"
    "        Từ xe tự lái đến trợ lý ảo, AI đã trở thành một phần không thể thiếu\n"
    "        trong cuộc sống hàng ngày. Các công ty công nghệ lớn đang đầu tư\n"
    "        hàng tỷ đô la vào nghiên cứu AI, với hy vọng tạo ra những đột phá\n"
    "        mới trong lĩnh vực này.\n"
    "        "
)


async def main() -> None:
    operonx.bootstrap()  # loads ./.env + ./resources.yaml from CWD

    runs = [
        ("basic", basic_chat(question="Python là gì? Trả lời trong 1 câu.")),
        ("summarize", summarize_pipeline(text=SAMPLE_TEXT)),
    ]
    for label, g in runs:
        result = await Operon(g).run(inputs={})
        # Strip $state for tidier output
        content = {k: v for k, v in result.items() if k != "$state"}
        print(f"[{label}] {content}")


if __name__ == "__main__":
    asyncio.run(main())

# ── the served front door ───────────────────────────────────────────────
# Every operonx project serves. The [[serve]] block in operonx.toml names
# this graph, `operonx-serve` boots it, and the studio draws it as the
# entry node feeding the flow — no pipeline begins from nowhere.
#
# `ingress` yields one item per request payload and `egress` writes the
# reply back to the caller. Neither names a resource: the run was minted
# by a transport and already carries its session — and with no session the
# same graph still runs under a plain `engine.start()`, so serving costs
# the example nothing.
from operonx.core.serve import egress, ingress


@op
def answer(item=None) -> dict:
    """One request in, this example's reply out."""
    return {"reply": f"ex03 saw: {item!r}"}


@graph
def served():
    request = ingress()
    a = answer(item=request["item"])
    out = egress(item=a["reply"])
    START >> request >> a >> out >> END

File: examples/rust/ex03_llm_chat/src/main.rs
//! 03 LLM Chat — Rust-side demo.
//!
//! Mirrors `examples/python/ex03_llm_chat/main.py`. `PromptOp` + `LLMOp`
//! are runtime-built provider ops materialised from `graph.json`; we
//! only declare the plain `#[op]`s the engine needs to resolve.
//!
//! Requires `OPENAI_API_KEY` in `.env` and `llm:gpt-4o-mini` in
//! `resources.yaml` (both shipped alongside this crate).
//!
//! ```sh
//! cargo run --release
//! ```

use operonx::{op, BootstrapOpts, Operon};
use serde_json::Value;

#[op(name = "clean_text")]
fn clean_text(text: String) -> Value {
    let cleaned: String = text.split_whitespace().collect::<Vec<_>>().join(" ");
    serde_json::json!({ "cleaned_text": cleaned })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let here = std::path::Path::new(env!("CARGO_MANIFEST_DIR"));

    // Load .env + resources.yaml from this crate's own directory so the
    // demo runs no matter where `cargo run` is invoked from.
    let opts = BootstrapOpts::new().resources(here.join("resources.yaml"));
    operonx::bootstrap(opts);

    let graph_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("graph.json"))?)?;
    let inputs_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("inputs.json"))?)?;

    for name in ["basic", "chain", "summarize"] {
        let graph_v = graph_bundle
            .get(name)
            .ok_or_else(|| format!("graph.json missing `{name}` entry"))?;
        let graph_json = serde_json::to_string(graph_v)?;

        let inputs_obj = inputs_bundle
            .get(name)
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();

        let engine = Operon::builder(&graph_json).auto_register().build()?;
        let result = engine.run_json(inputs_obj, None, None, None)?;
        println!("[{name}] {result}");
    }

    Ok(())
}
